From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mallikarjun Kasoju <mkasoju@nvidia.com>
Date: Mon, 27 Jan 2025 10:25:23 +0000
Subject: [PATCH 084/240] NVIDIA: SAUCE: scsi: ufs: skip ufs init if already
 initialised in earlier boot flow

Bug 4757621

Change-Id: I1b00c7f4b1d80bfb7d63802a6cfcbddab50e95f2
Signed-off-by: Mallikarjun Kasoju <mkasoju@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3318624
Reviewed-by: Bitan Biswas <bbiswas@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit c3c1db5b1103af75f0fb704ba2043c30a8c4a61c)

diff --git a/drivers/ufs/core/ufshcd.c b/drivers/ufs/core/ufshcd.c
index 6b2a096d9c8b..090361e06433 100644
--- a/drivers/ufs/core/ufshcd.c
+++ b/drivers/ufs/core/ufshcd.c
@@ -4674,6 +4674,11 @@ static int ufshcd_change_power_mode(struct ufs_hba *hba,
 {
 	int ret;
 
+
+	if (hba->is_ufs_already_enabled) {
+		dev_dbg(hba->dev, "%s: power already configured\n", __func__);
+		return 0;
+	}
 	/* if already configured to the requested pwr_mode */
 	if (!hba->force_pmc &&
 	    pwr_mode->gear_rx == hba->pwr_info.gear_rx &&
@@ -4936,6 +4941,44 @@ void ufshcd_hba_stop(struct ufs_hba *hba)
 }
 EXPORT_SYMBOL_GPL(ufshcd_hba_stop);
 
+static int ufshcd_hba_execute_hce_active(struct ufs_hba *hba)
+{
+	int retry_outer = 3;
+	int retry_inner;
+
+start:
+
+	ufshcd_set_link_off(hba);
+
+	ufshcd_vops_hce_enable_notify(hba, PRE_CHANGE);
+
+	ufshcd_delay_us(hba->vps->hba_enable_delay_us, 100);
+
+	/* wait for the host controller to complete initialization */
+	retry_inner = 50;
+	while (!ufshcd_is_hba_active(hba)) {
+		if (retry_inner) {
+			retry_inner--;
+		} else {
+			dev_err(hba->dev,
+				"Controller enable failed\n");
+			if (retry_outer) {
+				retry_outer--;
+				goto start;
+			}
+			return -EIO;
+		}
+		usleep_range(1000, 1100);
+	}
+
+	/* enable UIC related interrupts */
+	ufshcd_enable_intr(hba, UFSHCD_UIC_MASK);
+
+	ufshcd_vops_hce_enable_notify(hba, POST_CHANGE);
+
+	return 0;
+}
+
 /**
  * ufshcd_hba_execute_hce - initialize the controller
  * @hba: per adapter instance
@@ -4990,6 +5033,8 @@ static int ufshcd_hba_execute_hce(struct ufs_hba *hba)
 	ufshcd_enable_intr(hba, UFSHCD_UIC_MASK);
 
 	ufshcd_vops_hce_enable_notify(hba, POST_CHANGE);
+	usleep_range(1000, 1100);
+	ufshcd_clear_eh_in_progress(hba);
 
 	return 0;
 }
@@ -5018,9 +5063,11 @@ int ufshcd_hba_enable(struct ufs_hba *hba)
 
 		ufshcd_vops_hce_enable_notify(hba, POST_CHANGE);
 	} else {
-		ret = ufshcd_hba_execute_hce(hba);
+		if (hba->is_ufs_already_enabled)
+			ret = ufshcd_hba_execute_hce_active(hba);
+		else
+			ret = ufshcd_hba_execute_hce(hba);
 	}
-
 	return ret;
 }
 EXPORT_SYMBOL_GPL(ufshcd_hba_enable);
@@ -5086,7 +5133,7 @@ EXPORT_SYMBOL_GPL(ufshcd_update_evt_hist);
  */
 static int ufshcd_link_startup(struct ufs_hba *hba)
 {
-	int ret;
+	int ret = 0;
 	int retries = DME_LINKSTARTUP_RETRIES;
 	bool link_startup_again = false;
 
@@ -5102,7 +5149,8 @@ static int ufshcd_link_startup(struct ufs_hba *hba)
 	do {
 		ufshcd_vops_link_startup_notify(hba, PRE_CHANGE);
 
-		ret = ufshcd_dme_link_startup(hba);
+		if (!hba->is_ufs_already_enabled)
+			ret = ufshcd_dme_link_startup(hba);
 
 		/* check if device is detected by inter-connect layer */
 		if (!ret && !ufshcd_is_device_present(hba)) {
diff --git a/include/ufs/ufshcd.h b/include/ufs/ufshcd.h
index 0da0a0ed6804..73fb085c274d 100644
--- a/include/ufs/ufshcd.h
+++ b/include/ufs/ufshcd.h
@@ -1032,6 +1032,7 @@ struct ufs_hba {
 	u16 ee_usr_mask;
 	struct mutex ee_ctrl_mutex;
 	bool is_powered;
+	bool is_ufs_already_enabled;
 	bool shutting_down;
 	struct semaphore host_sem;
 
-- 
2.55.0

